test/compiler/inference.jl

```diff
@@ -1719,6 +1719,41 @@ for expr25261 in opt25261[i:end]
 end
 @test foundslot
 
+@testset "interprocedural conditional constraint propagation" begin
+    isaint(a) = isa(a, Int)
+    @test Base.return_types((Any,)) do a
+        isaint(a) && return a # a::Int
+        return 0
+    end == [Int]
+    eqnothing(a) = a === nothing
+    @test Base.return_types((Union{Nothing,Int},)) do a
+        eqnothing(a) && return 0
+        return a # a::Int
+    end == [Int]
+
+    # tests with base functions
+    @test Base.return_types((Any,)) do a
+        Base.Fix2(isa, Int)(a) && return sin(a) # a::Float64
+        return 0.0
+    end == [Float64]
+    @test Base.return_types((Union{Nothing,Int},)) do a
+        isnothing(a) && return 0
+        return a # a::Int
+    end == [Int]
+
+    # FIXME: we can't propagate conditional constraints interprocedurally when there're
+    # multiple possible conditions within the callee
+    ispositive(a) = isa(a, Int) && a > 0
```

Comment:
yes, I _think_ we can change our local inference logic so that we can annotate `Conditional` for a statement (e.g. `%5`) with considering where it comes from (e.g. `%1`), but I didn't try to tackle that in this PR because I felt the changes can be huge.
Maybe we can merge this PR with only introducing `InterConditional` and leave this as a future work (in a separate PR) ?

Reply:
The change should be pretty simple, though probably doesn't need to be in this PR: if the inferred result is <: Bool, then it is actually Conditional for the entire VarTable `s[pc][1:nargs]`. The annoying part then is that InterConditional can only represent exactly one of those args, so you may have to just pick the first interesting one. (conceptually, this is a also a generalizable observation on what Conditional represents: either it's an alternative VarTable based on a particular value comparison result—or the main VarTable `s` is a Conditional on the program state)

Reply:
Okay, with the following diff, we can back-propagate constraints in cases like `ispositive` and also some cases like 
the original definitions of `isnothing` and `ismissing` (it looks dirty, though):
```julia
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 98b8b1c9c9..6a09b74b3e 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -1300,6 +1300,7 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
     W = frame.ip
     s = frame.stmt_types
     n = frame.nstmts
+    nargs = frame.nargs
     while frame.pc´´ <= n
         # make progress on the active ip set
         local pc::Int = frame.pc´´ # current program-counter
@@ -1362,6 +1363,25 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             elseif isa(stmt, ReturnNode)
                 pc´ = n + 1
                 rt = abstract_eval_value(interp, stmt.val, s[pc], frame)
+                if isa(rt, Conditional) && !(1 ≤ slot_id(rt.var) ≤ nargs)
+                    # discard this `Conditional` imposed on non-call arguments,
+                    # since it's not interesting in inter-procedural context;
+                    # other ReturnNode statement may give constraints on any of call argument
+                    rt = widenconditional(rt)
+                end
+                if rt !== Bottom && !isa(rt, Conditional) && rt ⊑ Bool
+                    if isa(frame.bestguess, Conditional)
+                        # if the bestguess so far is already `Conditional`, try to convert
+                        # this `rt` into `Conditional` on the slot to avoid overapproximation
+                        # due to conflict of different slots
+                        rt = boolean_rt_to_conditional(rt, changes, slot_id(frame.bestguess.var))
+                    elseif nargs ≥ 1
+                        # pick up the first "interesting" slot, convert `rt` to its `Conditional`
+                        # TODO: this is very naive heuristic, ideally we want something like
+                        # alternative `VarTable` that keeps conditional constraints on all the variables
+                        rt = boolean_rt_to_conditional(rt, changes, nargs == 1 ? 1 : 2)
+                    end
+                end
                 if !isa(rt, Const) &&
                    !isa(rt, Type) &&
                    !isa(rt, PartialStruct) &&
@@ -1478,6 +1498,21 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
     nothing
 end
 
+function boolean_rt_to_conditional(@nospecialize(rt), state::VarTable, slot_id::Int)
+    varstate = state[slot_id]::VarState
+    typ = widenconditional(varstate.typ) # avoid nested conditional
+    if isa(rt, Const)
+        if rt.val === true
+            return Conditional(SlotNumber(slot_id), typ, Bottom)
+        elseif rt.val === false
+            return Conditional(SlotNumber(slot_id), Bottom, typ)
+        end
+    elseif rt === Bool
+        return Conditional(SlotNumber(slot_id), typ, typ)
+    end
+    return rt # when `rt` is Union{}, etc.
+end
+
 # make as much progress on `frame` as possible (by handling cycles)
 function typeinf_nocycle(interp::AbstractInterpreter, frame::InferenceState)
     typeinf_local(interp, frame)
```

But I couldn't succeed to bootstrap with this (with `fatal error in type inference (type bound)` error).. 
Do you have any pointer about where to fix ?

Reply:
On closer inspection, I think you're missing `bestguess_to_interprocedural` on some of the uses of `bestguess`. While you could grep each usage, I think it might be better to consolidate it here: (I thought the new assert should have caught that)

```diff
diff --git a/base/compiler/abstractinterpretation.jl b/base/compiler/abstractinterpretation.jl
index 98b8b1c9c9..c76c61e89b 100644
--- a/base/compiler/abstractinterpretation.jl
+++ b/base/compiler/abstractinterpretation.jl
@@ -1362,14 +1362,11 @@ function typeinf_local(interp::AbstractInterpreter, frame::InferenceState)
             elseif isa(stmt, ReturnNode)
                 pc´ = n + 1
                 rt = abstract_eval_value(interp, stmt.val, s[pc], frame)
-                if !isa(rt, Const) &&
-                   !isa(rt, Type) &&
-                   !isa(rt, PartialStruct) &&
-                   !isa(rt, Conditional)
-                    # only propagate information we know we can store
-                    # and is valid inter-procedurally
-                    # some of them will further be transformed by `bestguess_to_interprocedural`
-                    # in `finish(::InferenceState, ::AbstractInterpreter)`
+                # only propagate information we know we can store
+                # and is valid inter-procedurally
+                if isa(rt, Conditional)
+                    rt = bestguess_to_interprocedural(rt, frame.nargs)
+                elseif !isa(rt, Const) && !isa(rt, Type) && !isa(rt, PartialStruct)
                     rt = widenconst(rt)
                 end
                 if tchanged(rt, frame.bestguess)
diff --git a/base/compiler/typeinfer.jl b/base/compiler/typeinfer.jl
index 8a2d22b252..6cbe1fb5f0 100644
--- a/base/compiler/typeinfer.jl
+++ b/base/compiler/typeinfer.jl
@@ -416,11 +416,10 @@ function finish(me::InferenceState, interp::AbstractInterpreter)
         me.result.src = OptimizationState(me, OptimizationParams(interp), interp)
     end
     me.result.valid_worlds = me.valid_worlds
-    me.result.result = me.bestguess = bestguess_to_interprocedural(me.bestguess, me.nargs)
+    me.result.result = me.bestguess
     nothing
 end
 
-bestguess_to_interprocedural(@nospecialize(bestguess), _::Int) = bestguess
 function bestguess_to_interprocedural(bestguess::Conditional, nargs::Int)
     # convert `Conditional` into `InterConditional` only when it imposes constrains on
     # any of call argument
@@ -787,7 +786,7 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
     # return the current knowledge about this cycle
     frame = frame::InferenceState
     update_valid_age!(frame, caller)
-    return bestguess_to_interprocedural(frame.bestguess, frame.nargs), nothing
+    return frame.bestguess, nothing
 end
 
 #### entry points for inferring a MethodInstance given a type signature ####
```

Reply:
Okay, dug deeper and found it is due to missing handling of Vararg expansion in this PR. Here's a reduced repro. Note how the `x...` turned into a `x::Tuple{Symbol}` type description (which will fail at runtime, because actually we know that `x::Symbol` there)

```julia
julia> code_typed(mapfilter, (typeof((x...) -> !isa(x, Tuple{Integer})), typeof(Base.delete!), Base.Set{Union{Int64, Symbol}}, Base.Set{Union{Int64, Symbol}}) , optimize=false)                       
1-element Vector{Any}:                                                                             
 CodeInfo(                                                                                         
    @ REPL[8]:2 within `mapfilter'                                                                                                                                                                     
1 ─ %1  = a::Set{Union{Int64, Symbol}}                                                                                                                                                                 
│         (@_6 = Base.iterate(%1))::Union{Nothing, Tuple{Union{Int64, Symbol}, Int64}}             
│   %3  = (@_6 === nothing)::Bool                                                                  
│   %4  = Base.not_int(%3)::Bool                                                                   
└──       goto #6 if not %4                                                                        
2 ┄ %6  = @_6::Tuple{Union{Int64, Symbol}, Int64}::Tuple{Union{Int64, Symbol}, Int64}              
│         (x = Core.getfield(%6, 1))::Union{Int64, Symbol}                                                                                                                                             
│   %8  = Core.getfield(%6, 2)::Int64                                                              
│   @ REPL[8]:3 within `mapfilter'                                                                 
│   %9  = (pred)(x)::Bool                                                                          
└──       goto #4 if not %9                                                                        
3 ─       (f)(b, x::Tuple{Symbol})::Any                                                                                                                                                                
└──       goto #4                                                                                  
4 ┄       (@_6 = Base.iterate(%1, %8))::Union{Nothing, Tuple{Union{Int64, Symbol}, Int64}}         
│   %14 = (@_6 === nothing)::Bool                                                                                                                                                                      
│   %15 = Base.not_int(%14)::Bool                                                                  
└──       goto #6 if not %15                                                                                                                                                                           
5 ─       goto #2                                                                                  
6 ┄       return nothing                                                                           
) => Nothing                              

julia> s = Set{Union{Int,Symbol}}((:a, 1)); mapfilter((x...) -> !isa(x, Tuple{Integer}), delete!, s, s)
signal (11): Segmentation fault                                                         
```

(was hoping to be able to explain simply how I found this, but ended up being a bit convoluted. Mostly I ran under `rr` until exit, then ran backwards to `break jl_throw`, then back to `watch` on the value of $rip, then back to `eval_function` a couple times, until I could see the printing of `p jl_(src)` for the correct `p jl_(li)` function. I searched this for "fatal" and saw `Tuple{Int}`, which I could tell was bad, so then it was "just" a matter of guessing different IR inputs until one propagated the type wrong)

Reply:
Thanks so much for your help !
Unfortunately I'm on macOS and it seems to me that debugging w/ gdb is hard on that platform (rr seems even unavailable ...). 
I'll try to get Linux or docker environment to develop Julia, but for now I'd like to move forward this PR.

12e1e79 enhances this PR by eagerly converting boolean conditions into `(Inter)Conditional`s following your idea.
Now I succeed in bootstrapping but it seems that there're few inference accuracy regressions.

Reply:
Okay I just fixed the inference accuracy regression; there're some code that make good use of `Const(val::Bool)` to improve inference quality (e.g. [`ifelse` tfunc](https://github.com/JuliaLang/julia/blob/d46198d1bbc19165e45f0b0a7003253ccff0773b/base/compiler/tfuncs.jl#L215-L231)), but now we turn `Const(val::Bool)` return type into `Conditional`, and so we need to respect its constant-ness to keep accuracy.
As for `ifelse` one, I fixed that in fe6accd, but there may be other points that should be tweaked.

Once we've done all of them, I think this PR should be good to go.

Reply:
Yes, on macOS, lldb is better, but setting up a VM just to run `rr` in linux is also great (Can you tell `rr` is one of my favorite magic tools?)

I think we also might want to stick with using `Const(::Bool)` as the result object when the new type doesn't refine the old type of the Argument (e.g. except when `⊑(s[pc][i], s[1][i])`), then choose later (in `abstract_call_gf_by_type`) to form a tighter bound if union-splitting with `napplicable > 1`. Note, you can examine `fieldtype(sig, n)` to find out the type of the nth argument.